fix: Connecting to Datasource - Rehaul (#691)
appsmithorg/appsmith-docs: website/docs/core-concepts/connecting-to-data-sources/connecting-to-databases.md

1. On the **Explorer** tab, click the **+** icon next to Datasources.
2. Select the sample `users` PostgreSQL database.
3. Go to **Datasources** and click the **Active** tab. Select the **New API +** button next to the `users` sample database.
4. Write the query in the query editor and click the **Run** button to execute it. For example:

```sql
SELECT * FROM users;
```
The select query retrieves all the data from the users table. You can display the query results in a Table widget by [binding the data](/core-concepts/data-access-and-binding/displaying-data-read#displaying-data-in-a-widget) in the Table Data property.

## General notes

:::info
Appsmith encrypts all your datasource credentials and stores them securely. Appsmith also doesn't store any data returned from your datasources and acts only as a proxy layer to orchestrate the execution of Queries. As Appsmith is an open source framework, you can [deploy it on-premise](/getting-started/setup), and audit it to ensure none of your data leaves your VPC. For more information, see [Security](/product/security#security-measures-within-appsmith). 
:::

> 

Reviewer: ```suggestion
### Connection pooling
```

> Appsmith creates a new connection pool with the database server when you first connect the database to your application. All subsequent queries executed by Appsmith against your database then re-use this connection to ensure that your queries are executed at run-time. In the case that an idle connection is closed by the database server, Appsmith creates a new connection while executing the next query.

Reviewer 2: Should the first line be `new connection pool...` or `new connection` ?

Reviewer: I agree. Let's add a heading above this para called **Connection pooling** and and update the sentence as Sumit suggested,

> Appsmith creates a new connection pool with the database server when you first connect the database to your application. All subsequent queries executed by Appsmith against your database then re-use this connection to ensure that your queries are executed at run-time. In the case that an idle connection is closed by the database server, Appsmith creates a new connection while executing the next query.

Reviewer: ```suggestion
Appsmith creates a new connection with the database server when you first connect the database to your application. All subsequent queries executed by Appsmith against your database then re-use this connection to ensure that your queries are executed at run-time. If an idle connection is closed by the database server, Appsmith creates a new connection while executing the next query.
```

The maximum number of concurrent queries that can be run on a database with Appsmith is 5. If the application tries to execute more concurrent queries, you may receive an error message stating that the connection isn't available.

## Troubleshooting
If you are experiencing difficulties with connecting datasources in Appsmith, you can refer to the [Datasource troubleshooting guide](https://chat.openai.com/help-and-support/troubleshooting-guide/action-errors/datasource-errors) for assistance. If you need further support, you can reach out on [Discord](https://discord.com/invite/rBTTVJp) or ask questions on the [community forum](https://community.appsmith.com/).

## Further reading

At this point, you should know enough to start a project of your own and start playing around with datasources. The resources mentioned below can be useful when you need to learn new skills:

* [Building UI](/core-concepts/building-ui/)
* [Queries](/core-concepts/data-access-and-binding/querying-a-database/)
* [Data Access and Binding](/core-concepts/data-access-and-binding)

> * [Connect to a localhost database/ API](/advanced-concepts/more/how-to-work-with-local-apis-on-appsmith)

Reviewer 3: ```suggestion
* [Connect to a localhost database or an API](/advanced-concepts/more/how-to-work-with-local-apis-on-appsmith)
```

---

# Databases

Databases are an essential part of many applications and are used to store and manage data. They provide a structured way to organize, retrieve, and manipulate data, and are designed to be efficient and reliable. You can run queries to read and write data directly from the Appsmith editor. 

### Connecting to a database

> <VideoEmbed host="youtube" videoId="sJIxtXInV14" title="How to connect to a datasource" caption="How to connect to a datasource | Example"/>

Reviewer: please include the Test datasource action in the video. and maybe move the video to the bottom after the instructions? If there is an intro you can think of before the video, then do add it as directly starting with a video looks a bit odd

Author: moved video after steps

> :::tip

Reviewer: Keep this information as Important info - info callout and name it important . Also mention clearly that it is for self hosted instances.

Also sharing a useful link on how to decide whether a callout should be tip, info, note , etc.- https://paligo.net/docs/en/admonitions--notes,-warnings,-tips,-etc--.html

Shorten the callout. Its too verbose. and remove the "how to whitelist appsmith on AWS" part . This should be cross-reference on the AWS datasource page

Author: done

Before connecting to a data source, you must whitelist the IP address of the Appsmith deployment on your database instance or VPC

18.223.74.85 and 3.131.104.27 are the IP addresses of the Appsmith cloud instances that need to be whitelisted

This is a guide on how to whitelist appsmith on AWS.
:::

* On the **Explorer tab**, click the **+** sign next to **Datasources**. 

> * You’ll see a [list of Datasources](/reference/datasources/) that Appsmith can connect to.

Reviewer: only hyperlink Datasources. also club the next point with this one 

```suggestion
*  Choose a Database from the [Datasources](/reference/datasources/) that Appsmith can connect to.
```

Author: added

> * Choose a **Database/API**.

Reviewer: remove this . clubbed in one point above

Author: removed

> * Provide the **configuration details** required to connect to your database. You may need to contact your database administrator to [whitelist appsmith cloud](/learning-and-resources/how-to-guides/aws-whitelist).

Reviewer: please remove the hyperlink. as that is only a guide to whitelist AWS which can be provided in the documentation for AWS. Actually that line (You may need to contact your database administrator.....) is not required because we are already providing a note above

Author: removed

* Click **Test** to verify that Appsmith can connect to your database using the details you provided.
* Rename and save your datasource.

Once your datasource has been added successfully, a success pop-up will appear at the top. 

> ### Sample database

Author: done

> Appsmith offer dummy database that you can use to experiment with the platform before adding your real data. You can connect to two sample datasets:

Author: done

> 1. **Movies:** This is a sample [MongoDB](/reference/datasources/querying-mongodb/) database with a collection called movies.

Author: updated

2. **Users:** This is a sample [PostgreSQL](/reference/datasources/querying-postgres) database with a table called users.

> :::note

Author: updated

> The **data** in the **mock database** gets **reset** every **24 hours** and is **common** for **all users** so it may contain **some unexpected values**.

Author: done

:::

> To better understand how to use the sample database, lets look at an example:

Author: removed

> <VideoEmbed host="youtube" videoId="l_MNNa9mg0w" title="Using A Sample Database " caption="How to use sample database | Example"/>

Author: changed the video and updated the caption

* Click on **+** next to Datasources.

> * Select the **user** datasource.

Author: done

> * Now, click the **+** icon next to the **queries/js** and choose the user datasource.

Author: made some changes

> * Add your code in the body section and **run** your query. For example:

Author: done

```js
SELECT * FROM users;
```

> The preceding query gets all the users data. You can use a [table widget](/reference/widgets/table) to display your data. 

Author: done

> :::info

Author: changed to note

> Appsmith applications are **secure-by-default**. All sensitive credentials, such as database credentials, are encrypted with [**AES-256 encryption**](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard). You can read more about security in this [document](/product/security#security-measures-within-appsmith). 

Author: done

:::

> 

Reviewer: I noticed that the General Notes and Concurrent Queries sections are missing. Please keep them. We shouldn't delete any sections of information during rehaul. We can restructure them by moving them to the right location and rewrite them to give more clarity, but we shouldn't cut out any information

## What's next

At this point, you should know enough to start a project of your own and start playing around with datasources. The resources mentioned below can be useful when you need to learn new skills:

> * [Building UI](/core-concepts/building-ui/)

Author: added - Data Access and Binding to further reading

---

# Databases

Databases are an essential part of many applications and are used to store and manage data. They provide a structured way to organize, retrieve, and manipulate data, and are designed to be efficient and reliable. You can run queries to read and write data directly from the Appsmith editor. 

### Connecting to a database

:::tip
Before connecting to a data source, you must whitelist the IP address of the Appsmith deployment on your database instance or VPC

18.223.74.85 and 3.131.104.27 are the IP addresses of the Appsmith cloud instances that need to be whitelisted

This is a guide on how to whitelist appsmith on AWS.
:::

* On the **Explorer tab**, click the **+** sign next to **Datasources**. 
* Choose a Database from the [Datasources](/reference/datasources/) that Appsmith can connect to.
* Provide the **configuration details** required to connect to your database. 
* Click **Test** to verify that Appsmith can connect to your database using the details you provided.
* Rename and save your datasource.

> <VideoEmbed host="youtube" videoId="sJIxtXInV14" title="How to connect to a datasource" caption="How to connect to a datasource | Example"/>

Author: Can you recheck this because, after the update, the lines may not be the same

> <VideoEmbed host="youtube" videoId="sJIxtXInV14" title="How to connect to a datasource" caption="How to connect to a datasource | Example"/>

Reviewer 3: This is still open.

Once your datasource has been added successfully, a success pop-up appears at the top. 

### Mock database

Appsmith provides two sample datasets to help you connect the data and test the platform. 

1. The first database is a sample [MongoDB](/reference/datasources/querying-mongodb/) database with a collection called "movies"
2. The second database is a sample [PostgreSQL](/reference/datasources/querying-postgres) database with a table called "users"

The mock databases are public and shared by all users, which may result in mangled data. To refresh the data, the databases are automatically **reset every day**, so any updates made to these databases are temporary.

<VideoEmbed host="youtube" videoId="l_MNNa9mg0w" title="Using A Sample Database " caption="How to use sample database | Example"/>

* Click on **+** next to Datasources.
* Select the mock **users** PostgreSQL database.
* Now, click the **+** icon next to the **queries/js** and choose the user datasource.
* Add your code in the body section and **run** your query. For example:

> ```js

Author: done

SELECT * FROM users;
```

> The preceding query retrieves all the data from the users table. You can display the query results in a Table widget by [binding the data](/core-concepts/data-access-and-binding/displaying-data-read#displaying-data-in-a-widget) in the Table Data property.

Author: updated(there were 2 diff comment for same thing, so keeping @jnikhila 's version)

:::note

> Appsmith applications are **secure-by-default**. All sensitive credentials, such as database credentials, are encrypted with [**AES-256 encryption**](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard). Read more about [security](/product/security#security-measures-within-appsmith). 

Author: added after the intro

## General notes
Appsmith establishes a connection pool with the database server when you first connect the database to your app. This allows subsequent queries to be executed efficiently. If an idle connection is closed by the database server, Appsmith creates a new connection for the next query.

> For certain database plugins such as PostgreSQL and MySQL, Appsmith maintains a connection pool because multiple queries can't be run on a single connection.

Reviewer 2: (1) Should we mention other plugins too that have connection pool - Redis, MsSQL, Redshift ? 
(2) Can you please replace `maintains a connection pool` with `explicitly creates and maintains a connection pool` ? - since for few plugins that underlying driver might do it implicitly. 
(3) Can you please replace `because multiple queries can't be run on a single connection` with `because usually a single connection is not designed to handle multiple simultaneous queries`

Author: done

> The maximum number of connections in the pool is limited to 5. This limit applies solely to the number of connections in the pool and not to the number of concurrent queries.

Reviewer 2: Can we rephrase this para as : 
```
The maximum number of connections in a pool is limited to 5. Please note that max 5 connections in a pool does not mean that a max of 5 queries can be run concurrently using this pool. The number of concurrent queries that this pool can run is usually much higher. 
```

Author: done

---

# Databases

Databases are an essential part of many applications and are used to store and manage data. They provide a structured way to organize, retrieve, and manipulate data, and are designed to be efficient and reliable. You can run queries to read and write data directly from the Appsmith editor. 

> :::note

Reviewer 3: Move this note to the landing page- [connecting to data sources](https://appsmith-docs-git-database-rehaul-get-appsmith.vercel.app/core-concepts/connecting-to-data-sources).

Appsmith applications are **secure-by-default**. All sensitive credentials, such as database credentials, are encrypted with [**AES-256 encryption**](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard). For more information, see [security](/product/security#security-measures-within-appsmith). 
:::

### Connecting to a database

:::info important

> As a self-hosted Appsmith user, you must [whitelist](https://docs.appsmith.com/learning-and-resources/how-to-guides/aws-whitelist) the IP address of the Appsmith deployment on your database instance or VPC. `18.223.74.85` and `3.131.104.27` are the IP addresses of the Appsmith cloud instances that need to be whitelisted

Reviewer 3: ```suggestion
As a self-hosted Appsmith user, it is necessary to [safelist](https://docs.appsmith.com/learning-and-resources/how-to-guides/aws-whitelist) the IP addresses of the Appsmith deployment on your database instance or VPC. The IP addresses to be added to the safelist are `18.223.74.85` and `3.131.104.27`, which are the IP addresses of the Appsmith cloud instances.
```

* You can create queries by selecting the **New API +** button available on the **datasource** page.
* Write the query in the query editor and click the **Run** button to execute it. For example:

```sql
SELECT * FROM users;
```
The select query retrieves the data from the users' table. You can display the query results in a Table widget by [binding the data](/core-concepts/data-access-and-binding/displaying-data-read#displaying-data-in-a-widget) in the Table Data property.

## General notes
Appsmith establishes a connection pool with the database server when you first connect the database to your app. This allows subsequent queries to be executed efficiently. If an idle connection is closed by the database server, Appsmith creates a new connection for the next query.

For certain database plugins such as PostgreSQL and MySQL, Appsmith maintains a connection pool because multiple queries can't be run on a single connection.

The maximum number of concurrent queries that can be run on a database with Appsmith is 5. If the application tries to execute more concurrent queries, you may receive an error message stating that the connection isn't available.

## Troubleshooting

> Are you having trouble connecting datasources with Appsmith? check out the [Datasource troubleshooting guide](/help-and-support/troubleshooting-guide/action-errors/datasource-errors) or reach out on [Discord](https://discord.com/invite/rBTTVJp) to get support or ask questions on the [community forum](https://community.appsmith.com/).

Reviewer 3: ```suggestion
If you are experiencing difficulties with connecting datasources in Appsmith, you can refer to the [Datasource troubleshooting guide](https://chat.openai.com/help-and-support/troubleshooting-guide/action-errors/datasource-errors) for assistance. If you need further support, you can reach out on [Discord](https://discord.com/invite/rBTTVJp) or ask questions on the [community forum](https://community.appsmith.com/).
```

---

# Databases

Databases are an essential part of many applications and are used to store and manage data. They provide a structured way to organize, retrieve, and manipulate data, and are designed to be efficient and reliable. You can run queries to read and write data directly from the Appsmith editor. 

### Connecting to a database

:::info important

If you are a self-hosted Appsmith user, it's necessary to whitelist the IP addresses of the Appsmith deployment on your database instance or VPC. 18.223.74.85 and 3.131.104.27 are the IP addresses of the Appsmith cloud instances that need to be whitelisted.
:::

<VideoEmbed host="youtube" videoId="sJIxtXInV14" title="How to connect to a datasource" caption="How to connect to a datasource"/>

* On the **Explorer tab**, click the **+** sign next to **Datasources**. 
* Choose a Database from the [Datasources](/reference/datasources/) that Appsmith can connect to.
* Provide the **configuration details** required to connect to your database. 
* Click **Test** to verify that Appsmith can connect to your database using the details you provided.

> * Rename and save your datasource.

Reviewer 2: Can we please add a line stressing on the importance of clicking the save button here - i.e. unless the save button is clicked the datasource configuration is not updated. Many users get confused here thinking that since the `Test` button shows that datasource is valid they can just go ahead and create a query.

Author: done

Once your datasource has been added successfully, a success pop-up appears at the top. 

### Mock database

Appsmith provides two sample datasets to help you connect the data and test the platform. 

1. A sample MongoDB database with a collection called "movies."
2. A sample PostgreSQL database with a table called "users."

:::note
The mock databases are public and shared by all users, which may result in mangled data. To refresh the data, the databases are automatically **reset every day**, so any updates made to these databases are temporary.
:::

<VideoEmbed host="youtube" videoId="TrV8h_Dvhbg" title="Using A Sample Database " caption="How to use mock database"/>

* Click on **+** next to Datasources.
* Select the mock **users** PostgreSQL database.
* You can create queries by selecting the **New API +** button available on the **datasource** page.
* Write the query in the query editor and click the **Run** button to execute it. For example:

```sql
SELECT * FROM users;
```

> The select query retrieves the data from the users' table. You can display the query results in a Table widget by [binding the data](/core-concepts/data-access-and-binding/displaying-data-read#displaying-data-in-a-widget) in the Table Data property.

Author: done

:::note
Databases configured within one page of an app can be accessed by developers anywhere within the workspace.
:::

## General notes
Appsmith establishes a connection pool with the database server when you first connect the database to your app. This allows subsequent queries to be executed efficiently. If an idle connection is closed by the database server, Appsmith creates a new connection for the next query.

For certain database plugins such as PostgreSQL and MySQL, Appsmith maintains a connection pool because multiple queries can't be run on a single connection.

> The maximum number of concurrent queries that can be run on a database with Appsmith is 5. If the application tries to execute more concurrent queries, you may receive an error message stating that the connection isn't available.

Reviewer 2: This is not correct, please remove this. We can instead mention that we hold max 5 connections in the pool. The limit of `5` is a limit on the number of max connections in a pool. `5` connections in a pool does not translate to `5` max concurrent queries. The number of concurrent queries can be much larger.

Author: updated
